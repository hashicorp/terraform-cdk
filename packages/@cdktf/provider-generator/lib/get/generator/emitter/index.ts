// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
export * from "./attributes-emitter";
export * from "./resource-emitter";
export * from "./struct-emitter";
