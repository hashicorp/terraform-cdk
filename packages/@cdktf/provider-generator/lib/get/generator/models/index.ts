// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
export * from "./attribute-model";
export * from "./attribute-type-model";
export * from "./resource-model";
export * from "./scope";
export * from "./struct";
