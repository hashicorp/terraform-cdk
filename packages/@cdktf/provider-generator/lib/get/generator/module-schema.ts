// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
export interface ModuleSchema {
  name: string;
  empty?: boolean;
  inputs: Input[];
  outputs: Output[];
}

export interface Output {
  name: string;
  type: string;
}

export interface Input {
  name: string;
  type: string;
  description?: string;
  default?: string;
  required: boolean;
}
