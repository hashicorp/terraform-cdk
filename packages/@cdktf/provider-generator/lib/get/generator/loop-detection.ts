// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { AttributeModel } from "./models/attribute-model";

type AttributeIdentifier = [string, Set<string>];
const equalAttributeIdentifiers = (
  x: AttributeIdentifier,
  y: AttributeIdentifier
) =>
  x[0] === y[0] &&
  x[1].size === y[1].size &&
  [...x[1]].every((x) => y[1].has(x));

function typeStructure(model: AttributeModel) {
  if (!model.type.isComplex) {
    return model.type.storedClassType;
  }

  return `<complex:{${model.type.struct?.attributes
    .map((att) => att.name)
    .sort()}}${model.type.typeModelType}>`;
}

function getAttributeIdentifier(model: AttributeModel): AttributeIdentifier {
  return [
    model.terraformName,
    new Set(
      model.type.struct!.attributes.map((a) => `${a.name}:${typeStructure(a)}`)
    ),
  ];
}

/**
 * Detects recursive attribute schemas and returns an Object
 * describing what property should be linking to which class
 */
export function detectAttributeLoops(attributes: AttributeModel[]): {
  [loopEntrypoint: string]: string; // attribute path to be replaced -> attribute path of the type to replace with
} {
  // We aproximate a class as a set of attributes names
  // This is not a perfect approximation, but it's good enough for our purposes
  const redirects: Record<string, string> = {};

  function depthFirstSearch(
    attribute: AttributeModel,
    path: string[] = [],
    knownStructs: { [attributePath: string]: AttributeIdentifier } = {}
  ) {
    const name = attribute.terraformName;
    const struct = attribute.type.struct;
    if (!struct) {
      return;
    }

    const structIdentifier = getAttributeIdentifier(attribute);
    // Detect if we visited this already
    const visited = Object.entries(knownStructs).find(([, attrIdentifier]) =>
      equalAttributeIdentifiers(structIdentifier, attrIdentifier)
    );

    if (visited) {
      // We have a loop, we don't need to search deeper
      const toBeReplacedWith = visited[0];
      const toReplace = [...path, name].join(".");
      redirects[toReplace] = toBeReplacedWith;
    } else {
      // Search deeper
      struct.attributes.forEach((a) =>
        depthFirstSearch(a, [...path, name], {
          ...knownStructs,
          // We haven't visited this yet, add it to the list of known structs
          [[...path, name].join(".")]: structIdentifier,
        })
      );
    }
  }

  attributes.forEach((attr) => depthFirstSearch(attr));

  return redirects;
}
