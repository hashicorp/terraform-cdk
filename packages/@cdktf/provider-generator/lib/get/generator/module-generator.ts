// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { CodeMaker, toCamelCase } from "codemaker";
import { ConstructsMakerModuleTarget } from "../constructs-maker";
import { AttributeModel } from "./models";
import { sanitizedComment } from "./sanitized-comments";

export class ModuleGenerator {
  constructor(
    private readonly code: CodeMaker,
    private readonly targets: ConstructsMakerModuleTarget[]
  ) {
    this.code.indentation = 2;

    for (const target of this.targets) {
      this.emitSubmodule(target);
    }
  }

  private emitSubmodule(target: ConstructsMakerModuleTarget) {
    const spec = target.spec;

    if (!spec) {
      throw new Error(`missing spec for ${target.fqn}`);
    }

    this.code.openFile(target.fileName);

    this.code.line(`// generated by cdktf get`);
    this.code.line(`// ${target.source}`);

    this.code.line(
      `import { TerraformModule, TerraformModuleUserConfig } from 'cdktf';`
    );
    this.code.line(`import { Construct } from 'constructs';`);

    const baseName = this.code.toPascalCase(target.name.replace(/[-/.]/g, "_"));
    const configType = `${baseName}Config`;

    this.code.openBlock(
      `export interface ${configType} extends TerraformModuleUserConfig`
    );
    for (const input of spec.inputs) {
      const optional = input.required && input.default === undefined ? "" : "?";

      const comment = sanitizedComment(this.code);
      if (input.description) {
        comment.line(input.description);
      }
      if (input.default) {
        comment.line(input.default);
      }
      if (input.type.includes("map(")) {
        comment.line(
          `The property type contains a map, they have special handling, please see {@link cdk.tf/module-map-inputs the docs}`
        );
      }
      comment.end();

      this.code.line(
        `readonly ${AttributeModel.escapeName(
          toCamelCase(input.name)
        )}${optional}: ${parseType(input.type)};`
      );
    }
    this.code.closeBlock();

    // Add a link to the Terraform Registry if it is sourced from there
    // https://developer.hashicorp.com/terraform/language/modules/sources
    // Registry modules are referred to as <NAMESPACE>/<NAME>/<PROVIDER>
    // Other sources contain dots, e.g.
    // app.terraform.io/example-corp/k8s-cluster/azurerm (private registries)
    // github.com/hashicorp/example (Github)
    // git@github.com:hashicorp/example.git
    // bitbucket.org/hashicorp/terraform-consul-aws (Bitbucket)
    // ... and more
    // ../module and ./module (local paths)
    const isNonRegistryModule = target.source.includes(".");
    let registryPath;
    // Submodules also exist (e.g. terraform-aws-modules/vpc/aws//modules/vpc-endpoints)
    // And linking directly to them in the Registry requires including the version
    // like e.g. https://registry.terraform.io/modules/terraform-aws-modules/vpc/aws/latest/submodules/vpc-endpoints
    if (target.source.includes("//")) {
      registryPath = target.source.replace(
        "//modules",
        `/${target.version || "latest"}/submodules`
      );
      // terraform-aws-modules/vpc/aws//modules/vpc-endpoints
      // ->
      // terraform-aws-modules/vpc/aws/latest/submodules/vpc-endpoints
    } else {
      // not submodule specified, just append the version
      registryPath = `${target.source}/${target.version || "latest"}`;
    }

    const comment = sanitizedComment(this.code);
    comment.line(`Defines an ${baseName} based on a Terraform module`);
    comment.line(``);
    comment.line(
      isNonRegistryModule
        ? `Source at ${target.source}`
        : `Docs at Terraform Registry: {@link https://registry.terraform.io/modules/${registryPath} ${target.source}}`
    );
    comment.end();
    this.code.openBlock(`export class ${baseName} extends TerraformModule`);

    this.code.line(`private readonly inputs: { [name: string]: any } = { }`);

    const allOptional = spec.inputs.find((x) => x.required) ? "" : " = {}";

    this.code.open(
      `public constructor(scope: Construct, id: string, config: ${configType}${allOptional}) {`
    );
    this.code.open(`super(scope, id, {`);
    this.code.line("...config,");
    this.code.line(`source: '${target.source}',`);
    if (target.version) {
      this.code.line(`version: '${target.version}',`);
    }
    this.code.close(`});`);

    for (const input of spec.inputs) {
      const inputName = AttributeModel.escapeName(toCamelCase(input.name));
      this.code.line(`this.${inputName} = config.${inputName};`);
    }

    this.code.close(`}`); // ctor

    for (const input of spec.inputs) {
      const inputName = AttributeModel.escapeName(toCamelCase(input.name));
      const inputType =
        parseType(input.type) +
        (input.required && input.default === undefined ? "" : " | undefined");
      this.code.openBlock(`public get ${inputName}(): ${inputType}`);
      this.code.line(`return this.inputs['${input.name}'] as ${inputType};`);
      this.code.closeBlock();

      this.code.openBlock(`public set ${inputName}(value: ${inputType})`);
      this.code.line(`this.inputs['${input.name}'] = value;`);
      this.code.closeBlock();
    }

    for (const output of spec.outputs) {
      const outputName = toCamelCase(output.name);
      this.code.openBlock(`public get ${outputName}Output()`);
      this.code.line(`return this.getString('${output.name}')`);
      this.code.closeBlock();
    }

    this.code.openBlock(`protected synthesizeAttributes()`);
    this.code.line(`return this.inputs;`);
    this.code.closeBlock();

    this.code.closeBlock(); // class
    this.code.closeFile(target.fileName);
  }
}

function parseType(type: string) {
  if (type === "string") {
    return "string";
  }
  if (type === "number") {
    return "number";
  }
  if (type === "bool") {
    return "boolean";
  }
  if (type === "list") {
    return "string[]";
  }
  if (type === "map") {
    return "{ [key: string]: string }";
  }
  if (type === "tuple") {
    return "string[]";
  }
  if (type === "any") {
    return "any";
  }

  const complexType = parseComplexType(type);
  if (complexType) {
    return complexType;
  }

  throw new Error(`unknown type ${type}`);
}

function parseComplexType(type: string): string | undefined {
  const complex = /^(object|list|map|set|tuple)\(([\s\S]+)\)/;
  const match = complex.exec(type);
  if (!match) {
    return undefined;
  }

  const [, kind, innerType] = match;

  if (kind === "object") {
    return `any`;
  }

  if (kind === "list" || kind === "set") {
    return `${parseType(innerType)}[]`;
  }

  if (kind === "tuple") {
    return `any[]`;
  }

  if (kind === "map") {
    return `{ [key: string]: ${parseType(innerType)} }`;
  }

  throw new Error(`unexpected kind ${kind}`);
}
