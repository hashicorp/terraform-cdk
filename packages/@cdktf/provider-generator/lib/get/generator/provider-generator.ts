// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { CodeMaker, toCamelCase } from "codemaker";
import { LANGUAGES, logger, TerraformProviderConstraint } from "@cdktf/commons";
import {
  FQPN,
  parseFQPN,
  ProviderName,
  ProviderSchema,
} from "./provider-schema";
import { ResourceModel } from "./models";
import { ResourceParser } from "./resource-parser";
import { ResourceEmitter, StructEmitter } from "./emitter";
import {
  ConstructsMakerProviderTarget,
  ConstructsMakerTarget,
} from "../constructs-maker";

interface ProviderData {
  name: string;
  source: string;
  version: string;
}

const isMatching = (
  target: ConstructsMakerTarget,
  terraformSchemaName: string
): boolean => {
  if (target.isModule) return false;

  const elements = terraformSchemaName.split("/");

  if (elements.length === 1) {
    return target.source === terraformSchemaName;
  } else {
    const [hostname, scope, provider] = elements;

    if (!hostname || !scope || !provider) {
      throw new Error(`can't handle ${terraformSchemaName}`);
    }

    return target.name === provider;
  }
};

export interface ProviderConstraints {
  [fqn: string]: ProviderData;
}

export class TerraformProviderGenerator {
  private resourceParser = new ResourceParser();
  private resourceEmitter: ResourceEmitter;
  private structEmitter: StructEmitter;
  public versions: { [fqpn: string]: string | undefined } = {};

  constructor(
    private readonly code: CodeMaker,
    private readonly schema: ProviderSchema
  ) {
    this.code.indentation = 2;
    this.resourceEmitter = new ResourceEmitter(this.code);
    this.structEmitter = new StructEmitter(this.code);
  }

  private getProviderByConstraint(
    providerConstraint: ConstructsMakerTarget
  ): FQPN | undefined {
    return Object.keys(this.schema.provider_schemas || {}).find((fqpn) =>
      isMatching(providerConstraint, fqpn)
    ) as FQPN | undefined;
  }

  public generate(providerConstraint: ConstructsMakerTarget) {
    const fqpn = this.getProviderByConstraint(providerConstraint);
    if (!fqpn) {
      logger.debug(
        `Could not find provider constraint for ${providerConstraint} in schema: ${JSON.stringify(
          this.schema,
          null,
          2
        )}`
      );
      throw new Error(
        `Could not find provider with constraint ${providerConstraint}`
      );
    }

    const providerVersion = this.schema.provider_versions?.[fqpn];
    this.emitProvider(fqpn, providerVersion, providerConstraint);
    this.versions[fqpn] = providerVersion;
  }

  public generateAll() {
    for (const fqpn of Object.keys(this.schema.provider_schemas || {})) {
      this.generate(
        new ConstructsMakerProviderTarget(
          new TerraformProviderConstraint(fqpn),
          LANGUAGES[0]
        )
      );
    }
  }

  public async save(outdir: string) {
    await this.code.save(outdir);
  }

  public buildResourceModels(fqpn: FQPN): ResourceModel[] {
    const provider = this.schema.provider_schemas?.[fqpn];
    if (!provider) {
      throw new Error(`Can not find provider '${fqpn}' in schema`);
    }

    const resources = Object.entries(provider.resource_schemas || {}).map(
      ([type, resource]) =>
        this.resourceParser.parse(fqpn, type, resource, "resource")
    );

    const dataSources = Object.entries(provider.data_source_schemas || {}).map(
      ([type, resource]) =>
        this.resourceParser.parse(fqpn, `data_${type}`, resource, "data_source")
    );

    return ([] as ResourceModel[]).concat(...resources, ...dataSources);
  }

  public getClassNameForResource(terraformType: string) {
    return this.resourceParser.getClassNameForResource(terraformType);
  }

  public getNamespaceNameForResource(terraformType: string) {
    return this.resourceParser.getNamespaceNameForResource(terraformType);
  }

  private emitProvider(
    fqpn: FQPN,
    providerVersion?: string,
    constraint?: ConstructsMakerTarget
  ) {
    const { name } = parseFQPN(fqpn);
    const provider = this.schema.provider_schemas?.[fqpn];
    if (!provider) {
      throw new Error(`Can not find provider '${fqpn}' in schema`);
    }

    const files: string[] = [];
    this.buildResourceModels(fqpn).forEach((resourceModel) => {
      if (constraint) {
        resourceModel.providerVersionConstraint = constraint.version;
        resourceModel.terraformProviderSource = constraint.source;
      }
      resourceModel.providerVersion = providerVersion;

      if (resourceModel.structsRequireSharding) {
        files.push(this.emitResourceWithComplexStruct(resourceModel));
      } else {
        files.push(this.emitResource(resourceModel));
      }
      this.emitResourceReadme(resourceModel);
    });

    if (provider.provider) {
      const providerResource = this.resourceParser.parse(
        fqpn,
        `provider`,
        provider.provider,
        "provider"
      );
      if (constraint) {
        providerResource.providerVersionConstraint = constraint.version;
        providerResource.terraformProviderSource = constraint.source;
      }
      providerResource.providerVersion = providerVersion;
      files.push(this.emitResource(providerResource));
      this.emitResourceReadme(providerResource);
    }

    this.emitIndexFile(name, files);
    this.emitLazyIndexFile(name, files);
  }

  private emitResourceReadme(resource: ResourceModel): void {
    const filePath = `${resource.namespaceFolderPath}/README.md`;
    this.code.openFile(filePath);
    this.code.line(`# \`${resource.terraformType}\``);
    this.code.line();
    const type = resource.isProvider
      ? resource.provider
      : resource.terraformType;
    this.code.line(
      `Refer to the Terraform Registory for docs: [\`${type}\`](${resource.linkToDocs}).`
    );
    this.code.closeFile(filePath);
  }

  private emitIndexFile(provider: ProviderName, files: string[]): void {
    const folder = `providers/${provider}`;
    const filePath = `${folder}/index.ts`;
    this.code.openFile(filePath);
    this.code.line("// generated by cdktf get");
    for (const file of files) {
      const dirName = file.replace(`${folder}/`, "").replace("/index.ts", "");
      this.code.line(
        `export * as ${toCamelCase(dirName)} from './${dirName}';`
      );
    }
    this.code.line();
    this.code.closeFile(filePath);
  }

  private emitLazyIndexFile(provider: ProviderName, files: string[]): void {
    const folder = `providers/${provider}`;
    const filePath = `${folder}/lazy-index.ts`;
    this.code.openFile(filePath);
    this.code.line("// generated by cdktf get");
    for (const file of files) {
      const dirName = file.replace(`${folder}/`, "").replace("/index.ts", "");
      this.code.line(
        `Object.defineProperty(exports, '${toCamelCase(
          dirName
        )}', { get: function () { return require('./${dirName}'); } });`
      );
    }
    this.code.line();
    this.code.closeFile(filePath);
  }

  private emitResource(resource: ResourceModel): string {
    this.code.openFile(resource.filePath);
    this.emitFileHeader(resource);
    this.structEmitter.emit(resource);
    this.resourceEmitter.emit(resource);
    this.code.closeFile(resource.filePath);

    return resource.filePath;
  }

  private emitResourceWithComplexStruct(resource: ResourceModel) {
    const generatedFiles = [];

    // drop the last segment of the filepath
    const filePath = resource.filePath;
    this.code.openFile(filePath);
    this.code.line(`// generated from terraform resource schema`);
    this.code.line();

    if (resource.structsRequireSharding) {
      if (resource.referencedTypes.length > 0) {
        this.code.line(
          `import { ${resource.referencedTypes.join(", \n")}} from './${
            resource.structsFolderName
          }'`
        );
      }

      this.code.line(`export * from './${resource.structsFolderName}'`);

      resource.importStatements.forEach((statement) =>
        this.code.line(statement)
      );

      this.structEmitter.emitInterface(resource, resource.configStruct);
      this.resourceEmitter.emit(resource);

      this.code.closeFile(filePath);

      this.structEmitter.emit(resource);
      generatedFiles.push(resource.fileName);
      generatedFiles.push(resource.structsFolderName);
    } else {
      resource.importStatements.forEach((statement) =>
        this.code.line(statement)
      );

      this.structEmitter.emit(resource);
      this.resourceEmitter.emit(resource);
      this.code.closeFile(filePath);
      generatedFiles.push(resource.fileName);
    }

    return filePath;
  }

  private emitFileHeader(resource: ResourceModel) {
    this.code.line(`// ${resource.linkToDocs}`);
    this.code.line(`// generated from terraform resource schema`);
    this.code.line();
    resource.importStatements.forEach((statement) => this.code.line(statement));
    this.code.line();
    this.code.line("// Configuration");
    this.code.line();
  }
}
