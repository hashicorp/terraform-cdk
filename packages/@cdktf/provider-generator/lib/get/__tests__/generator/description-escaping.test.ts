// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { TerraformProviderGenerator } from "../../generator/provider-generator";
import { CodeMaker } from "codemaker";

test("broken attribute description comments", async () => {
  const code = new CodeMaker();
  const workdir = fs.mkdtempSync(
    path.join(os.tmpdir(), "description-escaping.test")
  );
  const spec = JSON.parse(
    fs.readFileSync(
      path.join(
        __dirname,
        "fixtures",
        "description-escaping.test.fixture.json"
      ),
      "utf-8"
    )
  );
  new TerraformProviderGenerator(code, spec).generateAll();
  await code.save(workdir);

  const output = fs.readFileSync(
    path.join(workdir, "providers/google/description-escaping.ts"),
    "utf-8"
  );
  expect(output).toMatchSnapshot();
});

// That popped up in the google-beta provider which transforms markdown fenced
// code blocks from the actual markdown docs to the provider schema (``` => ''')
// see
//  - https://raw.githubusercontent.com/hashicorp/terraform-provider-google-beta/4094e3bdd530ca853a046a6d938807f8a131c7c7/website/docs/r/iam_workload_identity_pool_provider.html.markdown
//  - https://github.com/hashicorp/terraform-provider-google-beta/blob/4094e3bdd530ca853a046a6d938807f8a131c7c7/google-beta/resource_iam_beta_workload_identity_pool_provider.go

test("malformed code blocks which break in python rst", async () => {
  const code = new CodeMaker();
  const workdir = fs.mkdtempSync(
    path.join(os.tmpdir(), "markdown-description-with-code-blocks.test")
  );
  const spec = JSON.parse(
    fs.readFileSync(
      path.join(
        __dirname,
        "fixtures",
        "markdown-description-with-code-blocks.test.fixture.json"
      ),
      "utf-8"
    )
  );
  new TerraformProviderGenerator(code, spec).generateAll();
  await code.save(workdir);

  const output = fs.readFileSync(
    path.join(workdir, "providers/aws/code-blocks.ts"),
    "utf-8"
  );
  expect(output).toMatchSnapshot();
});
