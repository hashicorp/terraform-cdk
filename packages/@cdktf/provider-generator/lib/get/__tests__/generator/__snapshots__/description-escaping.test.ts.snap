// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`broken attribute description comments 1`] = `
"// https://www.terraform.io/docs/providers/google/r/description_escaping.html
// generated from terraform resource schema

import { Construct } from 'constructs';
import * as cdktf from 'cdktf';

// Configuration

export interface DescriptionEscapingConfig extends cdktf.TerraformMetaArguments {
  /**
  * The resource name of the Cloud KMS CryptoKey to be used to protect access 
 to messages published on this topic. Your project's PubSub service account 
 ('service-{{PROJECT_NUMBER}}@gcp-sa-pubsub.iam.gserviceaccount.com') must have 
 'roles/cloudkms.cryptoKeyEncrypterDecrypter' to use this feature. 
 The expected format is 'projects/*\\\\/locations/*\\\\/keyRings/*\\\\/cryptoKeys/*'
  * 
  * Docs at Terraform Registry: {@link https://www.terraform.io/docs/providers/google/r/description_escaping.html#broken_comments DescriptionEscaping#broken_comments}
  */
  readonly brokenComments: boolean | cdktf.IResolvable;
}

/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/description_escaping.html description_escaping}
*/
export class DescriptionEscaping extends cdktf.TerraformResource {

  // =================
  // STATIC PROPERTIES
  // =================
  public static readonly tfResourceType: string = \\"description_escaping\\";

  // ===========
  // INITIALIZER
  // ===========

  /**
  * Create a new {@link https://www.terraform.io/docs/providers/google/r/description_escaping.html description_escaping} Resource
  *
  * @param scope The scope in which to define this construct
  * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
  * @param options DescriptionEscapingConfig
  */
  public constructor(scope: Construct, id: string, config: DescriptionEscapingConfig) {
    super(scope, id, {
      terraformResourceType: 'description_escaping',
      terraformGeneratorMetadata: {
        providerName: 'google'
      },
      provider: config.provider,
      dependsOn: config.dependsOn,
      count: config.count,
      lifecycle: config.lifecycle
    });
    this._brokenComments = config.brokenComments;
  }

  // ==========
  // ATTRIBUTES
  // ==========

  // broken_comments - computed: false, optional: false, required: true
  private _brokenComments?: boolean | cdktf.IResolvable; 
  public get brokenComments() {
    return this.getBooleanAttribute('broken_comments') as any;
  }
  public set brokenComments(value: boolean | cdktf.IResolvable) {
    this._brokenComments = value;
  }
  // Temporarily expose input value. Use with caution.
  public get brokenCommentsInput() {
    return this._brokenComments;
  }

  // =========
  // SYNTHESIS
  // =========

  protected synthesizeAttributes(): { [name: string]: any } {
    return {
      broken_comments: cdktf.booleanToTerraform(this._brokenComments),
    };
  }
}
"
`;

exports[`malformed code blocks which break in python rst 1`] = `
"// https://www.terraform.io/docs/providers/aws/r/code_blocks.html
// generated from terraform resource schema

import { Construct } from 'constructs';
import * as cdktf from 'cdktf';

// Configuration

export interface CodeBlocksConfig extends cdktf.TerraformMetaArguments {
  /**
  * Here comes a code block \`\`\`foo.bar\`\`\` and here is more text.
  * 
  * Docs at Terraform Registry: {@link https://www.terraform.io/docs/providers/aws/r/code_blocks.html#with_code_block CodeBlocks#with_code_block}
  */
  readonly withCodeBlock: boolean | cdktf.IResolvable;
}

/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/code_blocks.html code_blocks}
*/
export class CodeBlocks extends cdktf.TerraformResource {

  // =================
  // STATIC PROPERTIES
  // =================
  public static readonly tfResourceType: string = \\"code_blocks\\";

  // ===========
  // INITIALIZER
  // ===========

  /**
  * Create a new {@link https://www.terraform.io/docs/providers/aws/r/code_blocks.html code_blocks} Resource
  *
  * @param scope The scope in which to define this construct
  * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
  * @param options CodeBlocksConfig
  */
  public constructor(scope: Construct, id: string, config: CodeBlocksConfig) {
    super(scope, id, {
      terraformResourceType: 'code_blocks',
      terraformGeneratorMetadata: {
        providerName: 'aws'
      },
      provider: config.provider,
      dependsOn: config.dependsOn,
      count: config.count,
      lifecycle: config.lifecycle
    });
    this._withCodeBlock = config.withCodeBlock;
  }

  // ==========
  // ATTRIBUTES
  // ==========

  // with_code_block - computed: false, optional: false, required: true
  private _withCodeBlock?: boolean | cdktf.IResolvable; 
  public get withCodeBlock() {
    return this.getBooleanAttribute('with_code_block') as any;
  }
  public set withCodeBlock(value: boolean | cdktf.IResolvable) {
    this._withCodeBlock = value;
  }
  // Temporarily expose input value. Use with caution.
  public get withCodeBlockInput() {
    return this._withCodeBlock;
  }

  // =========
  // SYNTHESIS
  // =========

  protected synthesizeAttributes(): { [name: string]: any } {
    return {
      with_code_block: cdktf.booleanToTerraform(this._withCodeBlock),
    };
  }
}
"
`;
